loki/transform/transform_inline.py

```diff
@@ -183,3 +188,123 @@ def inline_elemental_functions(routine):
         if all(hasattr(s, 'type') and s.type.dtype in removed_functions for s in im.symbols):
             import_map[im] = None
     routine.spec = Transformer(import_map).visit(routine.spec)
+
+
+def inline_member_routine(routine, member):
+    """
+    Inline an individual member :any:`Subroutine` at source level.
+
+    This will replace all :any:`Call` objects to the specified
+    subroutine with an adjusted equivalent of the member routines'
+    body. For this, argument matching, including partial dimension
+    matching for array references is performed, and all
+    member-specific declarations are hoisted to the containing
+    :any:`Subroutine`.
+
+    Parameters
+    ----------
+    routine : :any:`Subroutine`
+        The subroutine in which to inline all calls to the member routine
+    member : :any:`Subroutine`
+        The contained member subroutine to be inlined in the parent
+    """
+    # pylint: disable=import-outside-toplevel,cyclic-import
+    from loki.transform import recursive_expression_map_update
+
+    def _map_unbound_dims(var, val):
+        """
+        Maps all unbound dimension ranges in the passed array value
+        ``val`` with the indices from the local variable ``var``. It
+        returns the re-mapped symbol.
+
+        For example, mapping the passed array ``m(:,j)`` to the local
+        expression ``a(i)`` yields ``m(i,j)``.
+        """
+        new_dimensions = list(val.dimensions)
+
+        indices = [index for index, dim in enumerate(val.dimensions) if isinstance(dim, sym.Range)]
+
+        for index, dim in enumerate(var.dimensions):
+            new_dimensions[indices[index]] = dim
+
+        return val.clone(dimensions=tuple(new_dimensions))
+
+    # Prevent shadowing of member variables by renaming them a priori
+    parent_variables = routine.variable_map
+    duplicate_locals = tuple(
+        v for v in member.variables
+        if v.name in parent_variables and v.name not in member._dummies
```

Comment:
Do we risk problems with case sensitive comparisons for `v.name not in member._dummies`? I think the latter uses all lower-case variables, so might be worth amending the shadowing tests with an upper-case declaration of a shadowing variable?

Reply:
Wow, good catch, this does indeed break a special corner case that we didn't test yet. I've pushed an extension to the shadowing test and a fix for this.